File: internal/cli/rpc_helper.go
package cli

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"net"
	"time"

	"github.com/gorilla/websocket"

	"github.com/PineappleBond/xyncra-server/pkg/client"
	"github.com/PineappleBond/xyncra-server/pkg/protocol"
)

// standaloneRPC sends a single JSON-RPC request over a fresh WebSocket
// connection and returns the raw response data (resp.Data from the server).
// This is the shared fallback for all RPC-backed CLI commands when the
// daemon IPC is unavailable (D-032).
func standaloneRPC(ctx context.Context, cliCtx *CLIContext, method string, params map[string]any) (json.RawMessage, error) {
	dialCtx, dialCancel := context.WithTimeout(ctx, 5*time.Second)
	defer dialCancel()

	url := cliCtx.ServerURLWithUser()
	ws, _, err := websocket.DefaultDialer.DialContext(dialCtx, url, nil)
	if err != nil {
		return nil, fmt.Errorf("%s: %w", cliCtx.ServerURL, err)
	}
	defer ws.Close()

	paramsBytes, err := json.Marshal(params)
	if err != nil {
		return nil, fmt.Errorf("standalone marshal params: %w", err)
	}

	reqData, err := json.Marshal(protocol.PackageDataRequest{
		ID:     "1",
		Method: method,
		Params: json.RawMessage(paramsBytes),
	})
	if err != nil {
		return nil, fmt.Errorf("standalone marshal request: %w", err)
	}

	pkg := protocol.Package{
		Version: 1,
		Type:    protocol.PackageTypeRequest,
		Data:    json.RawMessage(reqData),
	}

	if err := ws.SetWriteDeadline(time.Now().Add(5 * time.Second)); err != nil {
		return nil, fmt.Errorf("standalone set write deadline: %w", err)
	}
	if err := ws.WriteJSON(pkg); err != nil {
		return nil, fmt.Errorf("standalone write: %w", err)
	}

	if err := ws.SetReadDeadline(time.Now().Add(5 * time.Second)); err != nil {
		return nil, fmt.Errorf("standalone set read deadline: %w", err)
	}
	var respPkg protocol.Package
	if err := ws.ReadJSON(&respPkg); err != nil {
		var opErr net.Error
		if errors.As(err, &opErr) && opErr.Timeout() {
			return nil, fmt.Errorf("standalone read: server timed out")
		}
		return nil, fmt.Errorf("standalone read: %w", err)
	}

	var resp protocol.PackageDataResponse
	if err := json.Unmarshal(respPkg.Data, &resp); err != nil {
		return nil, fmt.Errorf("standalone unmarshal response: %w", err)
	}

	if resp.Code != protocol.ResponseCodeOK {
		return nil, &client.ClientError{Code: resp.Code, Message: resp.Msg}
	}

	return resp.Data, nil
}

File: demo/web/packages/xyncra-client-cli/src/rpc-helper.ts
/**
 * Standalone WebSocket RPC fallback for when the daemon IPC is unavailable.
 *
 * Mirrors Go internal/cli/rpc_helper.go.
 * Opens a fresh WebSocket connection, sends a PackageDataRequest, reads
 * the PackageDataResponse, and returns the response data.
 *
 * @module
 */

import WebSocket from 'ws';
import type { PackageDataRequest, PackageDataResponse, Package } from '@xyncra/protocol';
import { PackageType, ResponseCode } from '@xyncra/protocol';
import type { CLIContext } from './cli-context.js';

/**
 * Send a single JSON-RPC request over a fresh WebSocket connection.
 * This is the shared fallback for CLI commands when IPC is unavailable (D-032).
 */
export async function standaloneRPC(
  cliCtx: CLIContext,
  method: string,
  params: Record<string, unknown>,
): Promise<unknown> {
  const url = cliCtx.getServerURLWithUser();

  return new Promise<unknown>((resolve, reject) => {
    const ws = new WebSocket(url);
    let settled = false;

    const timer = setTimeout(() => {
      if (!settled) {
        settled = true;
        ws.close();
        reject(new Error(`${cliCtx.serverURL}: server timed out`));
      }
    }, 10000);

    ws.on('open', () => {
      const req: PackageDataRequest = {
        id: '1',
        method,
        params,
      };

      const pkg: Package = {
        version: 1,
        type: PackageType.Request,
        data: req,
      };

      ws.send(JSON.stringify(pkg));
    });

    ws.on('message', (data: WebSocket.Data) => {
      if (settled) return;
      clearTimeout(timer);
      settled = true;
      ws.close();

      try {
        const respPkg = JSON.parse(data.toString()) as Package;
        const resp = respPkg.data as PackageDataResponse;

        if (resp.code !== ResponseCode.OK) {
          reject(new Error(resp.msg || `RPC error code ${resp.code}`));
          return;
        }

        if (isMutationMethod(method)) {
          process.stderr.write(
            'Note: Operation succeeded on server. Local database will be updated when daemon starts.\n',
          );
          process.stderr.write(
            "Hint: Run 'xyncra-client listen --user-id <user>' to enable local queries.\n",
          );
        }

        resolve(resp.data);
      } catch (err) {
        reject(new Error(`standalone unmarshal response: ${(err as Error).message}`));
      }
    });

    ws.on('error', (err) => {
      if (!settled) {
        clearTimeout(timer);
        settled = true;
        reject(new Error(`${cliCtx.serverURL}: ${err.message}`));
      }
    });

    ws.on('close', () => {
      clearTimeout(timer);
      if (!settled) {
        settled = true;
        reject(new Error('standalone: connection closed before response'));
      }
    });
  });
}

/**
 * Returns true if the RPC method modifies server state.
 * Used to display a hint about local database sync.
 */
export function isMutationMethod(method: string): boolean {
  switch (method) {
    case 'send_message':
    case 'delete_message':
    case 'mark_as_read':
    case 'create_conversation':
    case 'delete_conversation':
    case 'restore_conversation':
      return true;
    default:
      return false;
  }
}
